artcommon/artcommonlib/konflux/konflux_db.py

```diff
@@ -36,6 +323,112 @@ def bind(self, record_cls: typing.Type[KonfluxRecord]):
         self.bq_client.bind(record_cls.TABLE_ID)
         self.record_cls = record_cls
 
+    async def _ensure_group_cached(self, group: str):
+        """
+        Lazy-load cache for group if not already loaded.
+
+        Automatically loads the last N days of builds for the group on first access.
+        Uses per-group async events to coordinate concurrent load attempts - only one
+        coroutine will actually load while others wait for it to complete.
+
+        :param group: Group name (e.g., 'openshift-4.18')
+        """
+        if not self.cache:
+            return
+
+        # Quick check - if already loaded, return immediately
+        if self.cache.is_group_loaded(group):
+            self.logger.debug(f"Cache already loaded for group '{group}'")
+            return
+
+        # Get or create an event for this group (thread-safe)
+        with KonfluxDb._cache_lock:
+            if group not in KonfluxDb._group_loading_events:
+                # Create new event for this group
+                event = asyncio.Event()
+                KonfluxDb._group_loading_events[group] = event
+                should_load = True
+            else:
+                # Another coroutine is already loading
+                event = KonfluxDb._group_loading_events[group]
+                should_load = False
+
+        if not should_load:
+            # Wait for the other coroutine to finish loading
+            self.logger.debug(f"Waiting for another coroutine to load group '{group}'...")
+            await event.wait()
+            self.logger.debug(f"Cache loaded for group '{group}' by another coroutine")
+            return
+
+        # We're the one who will load
+        try:
+            # Double-check after getting permission - might have been loaded while we waited
+            if self.cache.is_group_loaded(group):
+                self.logger.debug(f"Cache already loaded for group '{group}' (loaded while acquiring event)")
+                event.set()  # Signal any waiters
+                return
+
+            self.logger.info(f"Lazy-loading cache for group '{group}' (last {self.cache._cache_days} days)...")
+
+            # Build query for last N days of builds in this group
+            start_time = datetime.now(tz=timezone.utc) - timedelta(days=self.cache._cache_days)
+            where_clauses = [
+                Column('group', String) == group,
+                Column('start_time', DateTime) >= start_time,
+            ]
```

Comment:
Adding a `Column('outcome', String).in_(['success', 'failure']),` would ensure we don't cache pending builds, which we're not interested in. A quick test showed 
```
Cache loaded 4519 builds for group 'openshift-4.21' (total: 4519)
```
versus
```
Cache loaded 9140 builds for group 'openshift-4.21' (total: 9140)
```